### ISSUE TITLE
Make DiagNormal and friends less restrictive, allowing other real types than Float64

### ISSUE THREAD

**Issue by oschulz** [JuliaStats/Distributions.jl#1493]
Currently, we have

```
const DiagNormal = MvNormal{Float64, PDiagMat{Float64,Vector{Float64}}, Vector{Float64}}
```

This is quite restrictive, it results in

```julia
julia> d = MvNormal(fill(0.0f1, 5), Diagonal(fill(1.0f0, 5)))

julia> d.Σ isa PDiagMat
true

julia> d isa DiagNormal
false
```

We just ran into the same with dual numbers - a diagonal `MvNormal` of `Float64` is a `DiagNormal`, but an `MvNormal` of `ForwardDiff.Dual` of `Float64` is not.

This is relevant since some methods in "mvnormal.jl" are specialized for `DiagNormal`.

`IsoNormal` and `FullNormal` are also limited to Float64, currently. I guess `Real` could be used for all of them?

**Comment by devmotion** [comment]
This would make the types non-concrete, I think it might be better to define
```julia
const IsoNormal{T<:Real} = MvNormal{T,ScalMat{T},Vector{T}}
```
etc. Unfortunately, in any case it would be a breaking change (e.g., dispatches on `SomeType{IsoNormal}` would break).

**Comment by oschulz** [comment]
> Unfortunately, in any case it would be a breaking change

Argh, yes ...

So maybe we change the specializations in "mvnormal.jl" to use things like `MvNormal{<:Real,ScalMat{<:Real},Vector{<:Real}}` directly, for now?

**Comment by devmotion** [comment]
Yes, that could be done. Might be a bit more convenient though to define and work with the more general type aliases (`AbstractIsoNormal{T}` etc. maybe?) and define the existing aliases as special cases (eg `const IsoNormal = AbstractIsoNormal{Float64}`).

**Comment by oschulz** [comment]
Yes, that sounds good!
